File: CPP/No313.cpp
/**
 * Created by Xiaozhong on 2020/8/27.
 * Copyright (c) 2020/8/27 Xiaozhong. All rights reserved.
 */

#include "iostream"
#include "vector"

using namespace std;

class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int> &primes) {
        int len = primes.size();
        vector<int> dp(n, 0);
        dp[0] = 1;
        vector<int> index(len, 0);

        for (int i = 1; i < n; i++) {
            int min_num = INT32_MAX;
            // 找到下一个丑数
            for (int j = 0; j < len; j++) {
                min_num = min(min_num, dp[index[j]] * primes[j]);
            }
            dp[i] = min_num;
            // 更新 index
            for (int j = 0; j < len; j++) {
                if (min_num == dp[index[j]] * primes[j]) index[j]++;
            }
        }
        return dp[n - 1];
    }
};

int main() {
    vector<int> primes = {2, 7, 13, 19};
    Solution s;
    cout << s.nthSuperUglyNumber(1, primes);
}

File: Java/src/No313.java
/*
 * Copyright (c) 8/9/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import com.sun.jdi.IntegerType;

import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;
import java.util.TreeSet;

public class No313 {
    public int nthSuperUglyNumber(int n, int[] primes) {
        Set<Long> set = new HashSet<>();
        PriorityQueue<Long> pq = new PriorityQueue<>();
        set.add(1L);
        pq.add(1L);
        int uglyNumber = 0;
        for (int i = 0; i < n; ++i) {
            long curr = pq.poll();
            uglyNumber = (int) curr;
            for (int prime : primes) {
                long next = curr * prime;
                if (set.contains(next)) continue;
                pq.add(next);
                set.add(next);
            }
        }
        return uglyNumber;
    }

    public static void main(String[] args) {
        int[] primes = {2, 7, 13, 19};
        No313 no313 = new No313();
        System.out.println(no313.nthSuperUglyNumber(12, primes));
    }
}
